src/hotspot/os/aix/os_perf_aix.cpp

```diff
@@ -897,29 +575,64 @@ int CPUInformationInterface::cpu_information(CPUInformation& cpu_info) {
 }
 
 class NetworkPerformanceInterface::NetworkPerformance : public CHeapObj<mtInternal> {
-  friend class NetworkPerformanceInterface;
+  NONCOPYABLE(NetworkPerformance);
+
  private:
+  char* allocate_string(const char* str) const;
+
+  public:
   NetworkPerformance();
-  NONCOPYABLE(NetworkPerformance);
   bool initialize();
   ~NetworkPerformance();
   int network_utilization(NetworkInterface** network_interfaces) const;
 };
 
-NetworkPerformanceInterface::NetworkPerformance::NetworkPerformance() {
-
-}
+NetworkPerformanceInterface::NetworkPerformance::NetworkPerformance() {}
 
 bool NetworkPerformanceInterface::NetworkPerformance::initialize() {
-  return true;
+  return initialize_libperfstat();
 }
 
-NetworkPerformanceInterface::NetworkPerformance::~NetworkPerformance() {
-}
+NetworkPerformanceInterface::NetworkPerformance::~NetworkPerformance() {}
+
+int NetworkPerformanceInterface::NetworkPerformance::network_utilization(NetworkInterface** network_interfaces) const {
+  ResourceMark rm;
+  int n_records = 0;
+  perfstat_netinterface_t* net_stats;
+  perfstat_id_t name_holder;
+  int records_allocated = 0;
+
+  assert(network_interfaces != NULL, "network_interfaces is NULL");
+
+  *network_interfaces = NULL;
+  strncpy(name_holder.name , FIRST_NETINTERFACE, IDENTIFIER_LENGTH);
+
+  // calling perfstat_<subsystem>(NULL, NULL, _, 0) returns number of available records
+  if ((n_records = libperfstat::perfstat_netinterface(NULL, NULL, sizeof(perfstat_netinterface_t), 0)) <= 0) {
+    return OS_ERR;
+  }
 
-int NetworkPerformanceInterface::NetworkPerformance::network_utilization(NetworkInterface** network_interfaces) const
-{
-  return FUNCTIONALITY_NOT_IMPLEMENTED;
+  records_allocated = n_records;
+  net_stats = NEW_RESOURCE_ARRAY(perfstat_netinterface_t, records_allocated);
+
+  // populate net_stats && update the number of records that have been populated
+  n_records = libperfstat::perfstat_netinterface(&name_holder, net_stats, sizeof(perfstat_netinterface_t), n_records);
+
+  // check for error
+  if (n_records < 0) {
+    return OS_ERR;
+  }
+
+  for (int i = 0; i < n_records; i++) {
+    // Create new Network interface *with current head as next node*
+    NetworkInterface* new_interface = new NetworkInterface(net_stats[i].name,
+                                                           net_stats[i].ibytes,
+                                                           net_stats[i].obytes,
+                                                           *network_interfaces);
```

Comment:
When looking at NetworkInterface, I saw that it copies the name into its own resource-area allocated string in the constructor. This is all over the place, SystemProcess assumes C-heap storage, NetworkInterface uses RA... :(

The problem now is that your ResourceMark above may also destroy the internal string in NetworkInterface when the Stack is unwound beyond this function.

Side note: this is what I don't like about how hotspot uses RA. Its often not clear which data are RA allocated without looking at the code. In Objective-C (at least how its used on Mac) they had a clear naming convention for methods that returned storage from the AutoReleasePool.

The problem is also that with ResourceMarks cutting off your data, you may not see the problem until much later. Usually you only see immediate problems if the underlying arena (a chained list of malloced slabs) added a new slab under your ResourceMark, which then gets free'd when the ResourceMark goes out of scope.

How to fix this:
1) revert to NEW_C_HEAP_ARRAY for your records array here, with appropriate cleanup, and remove ResourceMark
3) Change NetworkInterface class to use C-heap instead.

I'd opt for (1) since it avoids discussions about changing shared code. I opened https://bugs.openjdk.java.net/browse/JDK-8280912 to change NetworkInterface, but don't wait for me.

Reply:
I made the change in (1) as recommended.